lib/filterx/func-str.c

```diff
@@ -0,0 +1,459 @@
+/*
+ * Copyright (c) 2024 Axoflow
+ * Copyright (c) 2024 Szilard Parrag
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "filterx/func-str.h"
+#include "filterx/object-primitive.h"
+#include "filterx/expr-literal.h"
+#include "filterx/expr-literal-generator.h"
+#include "filterx/filterx-eval.h"
+#include "filterx/object-extractor.h"
+#include "filterx/object-string.h"
+#include "filterx/object-list-interface.h"
+#include "scratch-buffers.h"
+
+#define FILTERX_FUNC_STARTSWITH_USAGE "Usage: startswith(string, prefix, ignorecase=true)" \
+"or startswith(string, [prefix_1, prefix_2, ..], ignorecase=true)"
+
+#define FILTERX_FUNC_ENDSWITH_USAGE "Usage: endswith(string, suffix, ignorecase=true)" \
+"or endswith(string, [suffix_1, suffix_2, ..], ignorecase=true)"
+
+#define FILTERX_FUNC_INCLUDES_USAGE "Usage: includes(string, substring, ignorecase=true)" \
+"or includes(string, [substring_1, substring_2, ..], ignorecase=true)"
+
+typedef struct _FilterXStringWithCache
+{
+  FilterXExpr *expr;
+  const gchar *str_value;
+  gsize str_len;
+} FilterXStringWithCache;
+
+typedef gboolean (*FilterXExprAffixProcessFunc)(const gchar *haystack, gsize haystack_len, const gchar *needle,
+                                                gsize needle_len);
+typedef struct _FilterXExprAffix
+{
+  FilterXFunction super;
+  gboolean ignore_case;
+  FilterXExpr *haystack;
+  struct
+  {
+    FilterXExpr *expr;
+    GPtrArray *cached_strings;
+  } needle;
+  FilterXExprAffixProcessFunc process;
+} FilterXExprAffix;
+
+static GString *
+_format_str_obj(FilterXObject *obj)
```

Comment:
All the format functions return a `GString *`, but your cache is a `gchar *` (or it is `const gchar *` right now, but should probably be a `gchar *`). This discrepancy will become obvious when you fix the `fill_cache()` method.

Reply:
That was intentional, so that `FilterXStringWithCache::free` won't be able to call `free()` on `ScratchBuffers` backed memory. Should I have just kept it as a `GString`?

Reply:
filterxstringwithcache should own the cache if it is available, right? it only gets filled init time once.